controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,286 @@
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: controllerAgentName}),
+
+		pLister:   kubePodInformer.Lister(),
+		pvLister:  kubePersistentVolumeInformer.Lister(),
+		pvcLister: kubePersistentVolumeClaimInformer.Lister(),
+
+		pStoreSynced:   kubePodInformer.Informer().HasSynced,
+		pvStoreSynced:  kubePersistentVolumeInformer.Informer().HasSynced,
+		pvcStoreSynced: kubePersistentVolumeClaimInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(EnhancedDefaultControllerRateLimiter(), "longhorn-kubernetes-pod"),
+	}
+
+	kubePodInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			curPod := cur.(*v1.Pod)
+			kc.enqueuePodChange(curPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+	})
+
+	return kc
+}
+
+func (kc *KubernetesPodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start %v", controllerAgentName)
+	if !controller.WaitForCacheSync(controllerAgentName, stopCh, kc.pStoreSynced, kc.pvStoreSynced, kc.pvcStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	logrus.Infof("Shutting down %v", controllerAgentName)
+}
+
+func (kc *KubernetesPodController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesPodController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesPodController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("%v: Error syncing Longhorn kubernetes pod %v: %v", controllerAgentName, key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("%v: Dropping Longhorn kubernetes pod %v out of the queue: %v", controllerAgentName, key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesPodController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", controllerAgentName, key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	pod, err := kc.pLister.Pods(namespace).Get(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Pod: %s", name)
+	}
+	nodeID := pod.Spec.NodeName
+	if err := kc.handlePodDeletionIfNodeDown(pod, nodeID, namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
```

Comment:
good commeting, have the first line be go doc compatible maybe like this:

// handlePodDeletionIfNodeDown determines whether we are allowed to forcefully delete a pod from a failed node based on the users chosen `NodeDownPodDeletionPolicy` this is necessary ...
// Because K8s ....

Reply:
Thank you! Fixed